thewalrus/quantum.py

```diff
@@ -1048,3 +1059,87 @@ def fock_tensor(S, alpha, cutoff, choi_r=np.arcsinh(1.0), check_symplectic=True,
+    mu and covariance matrix cov up to Fock space cutoff.
+
+    Args:
+        mu (array): vector of means of length 2*n_modes
+        cov (array): covariance matrix of shape [2*n_modes, 2*n_modes]
+        cutoff (int): Fock space cutoff
+        hbar (float): value of hbar
+
+    Returns:
+        (array): Fock space probabilities up to cutoff. The shape of this tensor is [cutoff]*num_modes
+    """
+    if is_pure_cov(cov, hbar=hbar):  # Check if the covariance matrix cov is pure
+        return np.abs(state_vector(mu, cov, cutoff=cutoff, hbar=hbar, check_purity=False)) ** 2
+    num_modes = len(mu) // 2
+    probabilities = np.zeros([cutoff] * num_modes)
+    for i in product(range(cutoff), repeat=num_modes):
+        probabilities[i] = np.maximum(
+            0.0, np.real_if_close(density_matrix_element(mu, cov, i, i, hbar=hbar))
+        )
+        # The maximum is needed because every now and then a probability is very close to zero from below.
+    return probabilities
+
+
+@jit(nopython=True)
+def loss_mat(eta, cutoff): # pragma: no cover
+    """ Constructs a binomial loss matrix with transmission eta up to n photons.
+
+    Args:
+        eta (float): Transmission coefficient. eta=0.0 means complete loss and eta=1.0 means no loss.
+        n (int): photon number cutoff.
+
+    Returns:
+        array: :math:`n\times n` matrix representing the loss.
+
+
+    """
+    # If full transmission return the identity
+
+    if eta < 0.0 or eta > 1.0:
+        raise ValueError("The transmission parameter eta should be a number between 0 and 1.")
+
+    if eta == 1.0:
+        return np.identity(cutoff)
+
+    # Otherwise construct the matrix elements recursively
+    lm = np.zeros((cutoff, cutoff))
+    mu = 1.0 - eta
+    lm[:, 0] = mu ** (np.arange(cutoff))
+    for i in range(cutoff):
+        for j in range(1, i + 1):
+            lm[i, j] = lm[i, j - 1] * (eta / mu) * (i - j + 1) / (j)
+    return lm
+
+
+def update_probabilities_with_loss(etas, probs):
+    """ Given a list of transmissivities etas and a tensor of probabilitites probs it calculates
+    an updated tensor of probabilities after loss is applied.
+
+    Args:
+        etas (list): List of transmission describing the loss in each of the modes
+        probs (array): Array of probabilitites in the different modes
+
+    Returns:
+        array: List of loss-updated probabilities.
+
+    """
+
+    probs_shape = probs.shape
+    if len(probs_shape) != len(etas):
+        raise ValueError("The list of transmission etas and the tensor of probabilities probs have incompatible dimensions.")
+
+    alphabet = "abcdefghijklmnopqrstuvwxyz"
+    cutoff = probs_shape[0]
+    for i, eta in enumerate(etas):
+        einstrings = "ij,{}i...->{}j...".format(alphabet[:i], alphabet[:i])
+
+        qein = np.zeros_like(probs)
+        qein = np.einsum(einstrings, loss_mat(eta, cutoff), probs)
+        probs = np.copy(qein)

```

Comment:
It does. Yes it is very unlikely anyone can ever do brute force Fock sampling with a reasonable cutoff for 26 modes. Yet, it would be nice to remove this restriction. Maybe @thisac could look into this?

Reply:
@josh146, how about we leave this for a future PR, or for whenever it is needed?

Reply:
Makes sense!